### PR TITLE
fix val_split logic and add tests

### LINKED ISSUE #203: val_split returns object of type 'NoneType' has no len()
## 🐛 Bug

When using the branch [`base_viz_2`](https://github.com/PyTorchLightning/lightning-flash/tree/base_viz_2) and I pass `val_split` to `ImageClassificationData.from_filepaths` I get the follow error:
```python
# 2. Load the data

datamodule = ImageClassificationData.from_filepaths(
    train_filepaths=train_images,
    train_labels=train_labels,
    batch_size=32,
    num_workers=mproc.cpu_count(),
    train_transform=TRAIN_TRANSFORM,
    val_transform=VALID_TRANSFORM,
    val_split=0.3  # crashes with base_viz_2
)
```
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-30-01c3b871d89f> in <module>
      8     train_transform=TRAIN_TRANSFORM,
      9     val_transform=VALID_TRANSFORM,
---> 10     val_split=0.3  # crashes with base_viz_2
     11 )

/opt/conda/lib/python3.7/site-packages/flash/vision/classification/data.py in from_filepaths(cls, train_filepaths, train_labels, val_filepaths, val_labels, test_filepaths, test_labels, predict_filepaths, train_transform, val_transform, test_transform, predict_transform, batch_size, num_workers, seed, preprocess_cls, **kwargs)
    529             preprocess=preprocess,
    530             seed=seed,
--> 531             **kwargs
    532         )

/opt/conda/lib/python3.7/site-packages/flash/data/data_module.py in from_load_data_inputs(cls, train_load_data_input, val_load_data_input, test_load_data_input, predict_load_data_input, preprocess, postprocess, **kwargs)
    375             data_pipeline = DataPipeline(
    376                 preprocess or cls(**kwargs).preprocess,
--> 377                 postprocess or cls(**kwargs).postprocess,
    378             )
    379         else:

/opt/conda/lib/python3.7/site-packages/pytorch_lightning/core/datamodule.py in __call__(cls, *args, **kwargs)
     48 
     49         # Get instance of LightningDataModule by mocking its __init__ via __call__
---> 50         obj = type.__call__(cls, *args, **kwargs)
     51 
     52         return obj

/opt/conda/lib/python3.7/site-packages/flash/vision/classification/data.py in __init__(self, train_dataset, val_dataset, test_dataset, predict_dataset, batch_size, num_workers, seed, train_split, val_split, test_split, **kwargs)
    214         if train_dataset is not None and train_split is not None or val_split is not None or test_split is not None:
    215             train_dataset, val_dataset, test_dataset = self.train_val_test_split(
--> 216                 train_dataset, train_split, val_split, test_split, seed
    217             )
    218 

/opt/conda/lib/python3.7/site-packages/flash/data/data_module.py in train_val_test_split(dataset, train_split, val_split, test_split, seed)
    294 
    295         """
--> 296         n = len(dataset)
    297 
    298         if test_split is None:

TypeError: object of type 'NoneType' has no len()
```

#### Code sample

https://colab.research.google.com/drive/1OHTFzWEock7lQ24SkD4oGDhdDkcU_mlf?usp=sharing

### Environment

In addition to that, in a Kaggle notebook environment the performance dropped in terms of 1-2min when I create an instance of `ImageClassificationData`.

### PR DESCRIPTION
**Pull request by edgarriba**
## What does this PR do?

Fixes #203 

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by kaushikb11** [comment]
@edgarriba Could you fix the tests?

**Review by tchaton — APPROVED** [review]
LGTM ! CI is failing.

**Comment by edgarriba** [comment]
@kaushikb11 @tchaton install packages seems to pull a broken version

File: requirements.txt [hunk 1]
```diff
@@ -17,3 +17,4 @@ sentencepiece>=0.1.95
 filelock  # comes with 3rd-party dependency
 pycocotools>=2.0.2 ; python_version >= "3.7"
 kornia>=0.5.0
+pytest-flake8
```
Comment (carmocca) [inline requirements.txt:20]: this is in the test requirements already

```suggestion
```

File: flash/vision/classification/data.py [hunk 2]
```diff
@@ -211,7 +211,7 @@ def __init__(
     ) -> 'ImageClassificationData':
         """Creates a ImageClassificationData object from lists of image filepaths and labels"""
 
-        if train_dataset is not None and train_split is not None or val_split is not None or test_split is not None:
+        if train_dataset is not None and train_split is not None or val_split is not None and test_split is not None:
```
Comment (edgarriba) [inline flash/vision/classification/data.py:214]: @tchaton I'm just noticing that this fix doesn't solve at al the issues. This prevents of not raising an error of `len(dataset)==0` but later the validation stage it's triggered.

File: requirements.txt (hunk 1 above)
Comment (edgarriba) [reply requirements.txt:20]: as a new contributor if you follow the contributions guidelines
https://github.com/PyTorchLightning/lightning-flash/blob/master/.github/CONTRIBUTING.md#test-cases

you will call `make test` which without `pytest-flake8` you won't succeed with detecting the CI failures.
even with that the CI calls another version of flake8 and raises more errors than the ones you get in local dev_env

/cc @aribornstein

Comment (carmocca) [reply requirements.txt:20]: In that case, we should mention in CONTRIBUTING to do `pip install .[dev,test]` as we do in Lightning before running `make test`.

https://github.com/PyTorchLightning/pytorch-lightning/blob/55525031c65756c2e8574cfee5d386e584aecdca/setup.py#L36-L48

and iron out the inconsistencies between the CI and local environment

I strongly suggest leaving these changes for a separate PR

cc: @Borda

File: flash/data/data_module.py [hunk 3]
```diff
@@ -374,8 +374,15 @@ def from_load_data_inputs(
...
-                postprocess or cls(**kwargs).postprocess,
+                preprocess, postprocess
+                # TODO: verify why the combination of paremeters below are needed and add tests.
+                # Otherwise there is a deadlock due to cyclic parameters dependencies. When we
+                # pass `val_split` since the the DataModule class expects a dataset but at this
+                # point is not still created. On the opther hand, to generate_dataset_if_possible
+                # needs the datapipeline.
+                ####
+                # preprocess or cls(**kwargs).preprocess,
+                # postprocess or cls(**kwargs).postprocess,
```
Comment (edgarriba) [inline flash/data/data_module.py:385]: @tchaton mind checking this

**Pull request closed without merging**